csrc/fusion_profiler.h

```diff
@@ -0,0 +1,319 @@
+  bool cupti_disabled_;
+
+  int device_;
+  uint32_t segment_id_;
+
+  HostTimer compile_timer_;
+  int64_t input_bytes_;
+  int64_t output_bytes_;
+  ProfilerState kernel_profile_state_;
+};
+
+//! \struct FusionProfiler
+//! \brief A singleton class to profile Fusions that can include multiple
+//! segments.
+class FusionProfiler {
+  FusionProfiler();
+
+ public:
+  //! Method to access FusionProfiler singleton
+  static FusionProfiler* get();
+
+  void reset();
+  void disableCupti(bool value);
+  ProfilerState state() const;
+  SegmentProfiler& segment(size_t idx);
+
+  //! Profiling Methods
+  void start();
+  void stop();
+  void createSegments(size_t num);
+  void startParallelCompile();
+  void stopParallelCompile();
+  void inputBytesAccessed(int64_t bytes);
+  void outputBytesAccessed(int64_t bytes);
+  const FusionProfile& profile() const;
+
+  //! Methods to capture Asynchronous CUPTI activity that get called from
+  //! functions registered with CUPTI.
+  //! Correlation ID -> Segment ID
+  void recordAsyncCorrIdActivity(uint32_t seg_id, uint32_t corr_id);
+  //! Collects CUPTI Kernel Activity
+  void recordAsyncKernelActivity(KernelProfile prof);
+  //! Ptr to the CUPTI Activity Buffer
+  uint8_t* cuptiBufferPtr();
+
+ private:
+  static FusionProfiler* singleton_;
+  static std::mutex singleton_lock_;
+
+  //! Disables CUPTI usage in order to measure Host Time without CUPTI overhead
+  bool cupti_disabled_;
+  //! Buffer for Cupti to store Activity Buffers during async activity
+  std::vector<uint8_t> cupti_buffer_;
+  //! The state is used to check for errors in usage
+  ProfilerState state_;
+
+  //! Data members with information that is aggregated into a FusionProfile
+  int64_t fusion_id_;
+  bool parallel_compile_;
+  FusionProfile profile_;
+  CudaEventTimer fusion_timer_;
+  HostTimer host_timer_;
+  HostTimer parallel_compile_timer_;
+  std::vector<SegmentProfiler> segments_;
+  //! The FusionProfiler collects a cache of device descriptors so each segment
+  //! does not need to spend time re-generating the information.
+  std::vector<DeviceDescriptor> device_descriptors_;
+
+  //! These 2 data members are used to collect and connect asynchronous records,
+  //! generated by CUPTI, to the segments responsible for the activity
+  std::vector<KernelProfile> kernel_profiles_;
+  std::unordered_map<uint32_t, uint32_t> corrid_2_segid_;
+};
+
+//! Macro to determine if the FusinProfiler is enabled
+#define _FP_ENABLE(code)                                            \
+  if (isDebugDumpEnabled(DebugDumpOption::FusionProfiler) ||        \
+      isDebugDumpEnabled(DebugDumpOption::FusionProfilerNocupti) || \
+      isDebugDumpEnabled(DebugDumpOption::FusionProfilerVerbose) || \
+      isOptionEnabled(EnableOption::FusionProfiler) ||              \
+      isOptionEnabled(EnableOption::FusionProfilerNocupti)) {       \
+    code;                                                           \
+  }
+
+//! Fusion Profiling Macros
```

Comment:
Is there a good reason to not use functions instead?

Reply:
Agree with Jingyue here.

It doesn't feel like the macro is saving us much code.

Reply:
Changed to functions.